src/modeling_library/mixture.jl

```diff
@@ -0,0 +1,283 @@
+            x_grad = if has_output_grad($base_dist)
+                sum(log_grads_x .* relative_weighted_densities)
+            else
+                nothing
+            end
+
+            # compute gradients with respect to the weights
+            weights_grad = exp.(log_densities .- logsumexp(log_weighted_densities))
+
+            # compute gradients with respect to each argument
+            $(Expr(:block, [quote
+                if has_argument_grads($base_dist)[$i]
+                    $(if dims[i] == 0
+                        # scalar argument
+                        quote
+                            grads = [log_grad[$(i+1)] for log_grad in log_grads]
+                            grad_weights = relative_weighted_densities
+                        end
+                    else
+                        # array argument
+                        quote
+                            grads = cat([log_grad[$(i+1)] for log_grad in log_grads]..., dims=$(dims[i] + 1))
+                            grad_weights = reshape(relative_weighted_densities, $((1 for d in 1:dims[i])..., length(dims)))
+                        end
+                    end)
+                    $(Symbol("$(arg)_grad")) = grads .* grad_weights
+                else
+                    $(Symbol("$(arg)_grad")) = nothing
+                end
+            end for (i, arg) in enumerate(args)]...))
+
+            return $(Expr(:tuple,
+                :x_grad, :weights_grad,
+                [:($(Symbol("$(arg)_grad"))) for arg in args]...))
+        end
+
+        # define callable behavior to be sampling
+        (::$(esc(name)))(weights, $(args...)) = random($(esc(name))(), weights, $(args...))
+
+        # evaluate to a singleton
+        $(esc(name))()
+
+    end # quote
+
+end # macro hom_mixture
+
+export @hom_mixture
+
+
+##############################################################################
+# heterogeneous mixture: fixed number of potentially different distributions #
+##############################################################################
+
+const MIXED_SYNTAX_ERR_MSG ="Expected syntax of the form @mixed Float64 (normal, 2) (uniform, 2)"
+const MIXED_WEIGHT_VECTOR_ERR_MSG = "length of weight vector must match the number of components"
+
+"""
+    @het_mixture output_type (base_dist1, num_args1) (base_dist2, num_args2), ...
+
+Defines a new type for a mixture distribution family, and evaluates to a
+singleton element of that type.
+
+The first argument, `output_type`, should match the output type of all of the
+base distributions.
+
+The remainder of arguments are tuples, where the first element must have type
+Gen.Distribution, and defines the distribution family of this component in the
+mixture, and the second element is a literal `Int` that gives the the number of
+arguments taken by this distribution family.
+
+Example:
+
+    uniform_beta_mixture = @het_mixture Float64 (uniform, 2) (beta, 2)
+"""
```

Comment:
Can we clarify here too that the resulting calling pattern is `dist(weights, args_for_dist_1..., args_for_dist_2...)`?